### PR TITLE
[Enhancement] extend query trace

### PR DESCRIPTION
**Pull request by dirtysalt**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

This PR is to extend query trace by:
1. add thread id to each event(so in theory we can build trace graph in POV of threads)
2. ~~add unique id to chunk source.~~ use chunk_source pointer as unique id
3. add pipeline driver name information.
4. output some fields in hex format(and change default id = 0 to make output shorter)

---

With following python script, we can change POV from pipeline driver to threads.

```python
#!/usr/bin/env python
# coding:utf-8
# Copyright (C) dirlt
import json
import sys


def load_events(fh):
    for s in fh:
        s = s.strip()
        if not s: continue
        if s.startswith('{'):
            if s.startswith('{"traceEvents"'): continue
            if s.startswith('{}'): continue
            if s.endswith(','):
                s = s[:-1]
            yield json.loads(s)


def remove_fields(d):
    for f in ('id', 'tidx', 'args'):
        if f in d:
            del d[f]


def fix_events(fh, events, prefilters, postfilters):
    io_tid_set = set()
    exec_tid_set = set()

    fh.write('{"traceEvents":[\n')

    # pid: fragment instance id low bits.
    # tid: pipeline driver pointer.
    def fix_io_task(ev):
        d = ev.copy()
        d.update({'pid': '0', 'tid': d['tidx'], 'name': d['cat']})
        d['ph'] = d['ph'].upper()
        d['cat'] = d['tid']
        # id: chunk source pointer.
        return d

    def fix_driver(ev):
        d = ev.copy()
        d.update({'pid': 0, 'tid': d['tidx'], 'name': d['cat'] + '_' + d['tid']})
        d['cat'] = d['tid']
        return d

    def fix_operator(ev):
        d = ev.copy()
        d.update({'pid': 0, 'tid': d['tidx']})
        return d

    for ev in events:
        name = ev.get('name')
        cat = ev.get('cat')
        d = None
        if any(f(ev) for f in prefilters): continue
        if name == 'io_task':
            d = fix_io_task(ev)
            io_tid_set.add(d['tid'])
        else:
            if name == 'process' and cat.startswith('driver'):
                d = fix_driver(ev)
            elif cat in ('pull_chunk', 'push_chunk'):
                d = fix_operator(ev)
            if d:
                exec_tid_set.add(d['tid'])
        if not d:
            continue
        remove_fields(d)
        if any(f(ev) for f in postfilters): continue
        fh.write(json.dumps(d) + ',\n')

    headers = []
    headers.append(dict(name='process_name', ph='M', pid=0, args={'name': 'starrocks_be'}))
    for tid in io_tid_set:
        d = dict(name='thread_name', ph="M", pid=0, tid=tid, args={"name": "IOThread-%s" % tid})
        headers.append(d)
    for tid in exec_tid_set:
        d = dict(name='thread_name', ph="M", pid=0, tid=tid, args={"name": "ExecThread-%s" % tid})
        headers.append(d)
    for d in headers:
        fh.write(json.dumps(d) + ',\n')
    fh.write('{}]}')


def drill_events(fh, events, text):
    fh.write('{"traceEvents":[\n')
    for ev in events:
        if ev['ph'] == 'M' or ev['name'].find(text) != -1:
            fh.write(json.dumps(ev) + ',\n')
    fh.write('{}]}')


def main():
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument('--input', help='input file')
    parser.add_argument('--output', help='output file')
    parser.add_argument('-v', action='store_true', help='output detailed version')
    parser.add_argument('-c', action='store_true', help='compress output file')
    parser.add_argument('--drill')
    args = parser.parse_args()

    fin = sys.stdin
    if args.input:
        fin = open(args.input)
    fout = sys.stdout
    if args.output:
        fout = open(args.output, 'w')

    prefilters = []
    postfilters = []
    if not args.v:
        prefilters.append(lambda x: x.get('cat') in ('push_chunk', 'pull_chunk'))

    events = load_events(fin)

    if args.drill:
        drill_events(fout, events, args.drill)
    else:
        fix_events(fout, events, prefilters, postfilters)


if __name__ == '__main__':
    main()

```

image

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

### CODE REVIEW

File: be/src/exec/pipeline/pipeline_driver.h [hunk 1]
```diff
@@ -158,6 +158,7 @@ class PipelineDriver {
         for (auto& op : _operators) {
             _operator_stages[op->get_id()] = OperatorStage::INIT;
         }
+        snprintf(_driver_name, sizeof(_driver_name), "Driver(node=%d/id=%d)", _source_node_id, _driver_id);
```
Comment (mofeiatwork) [inline be/src/exec/pipeline/pipeline_driver.h:161]: Why not a `std::string`? `snprintf` is so old school.

File: be/src/exec/pipeline/scan/scan_operator.cpp [hunk 2]
```diff
@@ -323,10 +323,10 @@ Status ScanOperator::_trigger_next_scan(RuntimeState* state, int chunk_source_in
             SCOPED_SET_TRACE_INFO(driver_id, state->query_id(), state->fragment_instance_id());
             SCOPED_THREAD_LOCAL_MEM_TRACKER_SETTER(state->instance_mem_tracker());
 
-            [[maybe_unused]] std::string category = "chunk_source_" + std::to_string(chunk_source_index);
-            QUERY_TRACE_ASYNC_START("io_task", category, query_trace_ctx);
-
             auto& chunk_source = _chunk_sources[chunk_source_index];
+            [[maybe_unused]] char category[64];
+            snprintf(category, sizeof(category), "chunk_source_0x%x", chunk_source->get_pseudo_unique_id());
```
Comment (mofeiatwork) [inline be/src/exec/pipeline/scan/scan_operator.cpp:328]: Same as above.

Comment (dirtysalt) [reply be/src/exec/pipeline/scan/scan_operator.cpp]: Fixed.

File: be/src/exec/pipeline/pipeline_driver.h (hunk 1 above)
Comment (dirtysalt) [reply be/src/exec/pipeline/pipeline_driver.h]: Fixed.

**Review by murphyatwork — DISMISSED** [review]

**Comment by silverbullet233** [comment]
could you add your python script into our repo so that others can use it easily?

**Comment by dirtysalt** [comment]
> could you add your python script into our repo so that others can use it easily?

good suggestion. I'll do that.

**Review by silverbullet233 — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Comment by dirtysalt** [comment]
run starrocks_admit_test

**Review by satanson — APPROVED** [review]

**Pull request merged**